reahl-component/reahl/component_dev/test_dbutils.py

```diff
@@ -57,20 +59,88 @@ def test_finding_database_control(dbcontrol_fixture):
         SystemControl(fixture.config)
 
 
-@with_fixtures(DBControlFixture)
-def test_database_control_settings(dbcontrol_fixture):
+class URIScenarios(Fixture):
+
+    @scenario
+    def all_settings_given(self):
+        self.uri = 'myprefix://theuser:thepasswd@thehost:123/thedb'
+        self.database_name = 'thedb'
+        self.user_name = 'theuser'
+        self.password = 'thepasswd'
+        self.host = 'thehost'
+        self.port = 123
+
+    @scenario
+    def not_all_settings_given(self):
+        self.uri = 'myprefix://theuser@thehost/thedb'
+        self.database_name = 'thedb'
+        self.user_name = 'theuser'
+        self.password = None
+        self.host = 'thehost'
+        self.port = None
+
+    @scenario
+    def minimal(self):
+        self.uri = 'myprefix:///:memory:'
+        self.database_name = ':memory:'
+        self.user_name = None
+        self.password = None
+        self.host = None
+        self.port = None
+
+
+@with_fixtures(DBControlFixture, URIScenarios)
+def test_database_control_settings(dbcontrol_fixture, uri_fixture):
     """DatabaseControl settings are read from the
        reahlsystem.connection_uri config setting parsed as an RFC1808 URI
     """
     fixture = dbcontrol_fixture
-    fixture.config.reahlsystem.connection_uri = 'myprefix://theuser:thepasswd@thehost:123/thedb'
+    fixture.config.reahlsystem.connection_uri = uri_fixture.uri
+    system_control = SystemControl(fixture.config)
+
+    assert system_control.db_control.database_name == uri_fixture.database_name
+    assert system_control.db_control.user_name == uri_fixture.user_name
+    assert system_control.db_control.password == uri_fixture.password
+    assert system_control.db_control.host == uri_fixture.host
+    assert system_control.db_control.port == uri_fixture.port
+
+
+@with_fixtures(DBControlFixture)
+def test_invalid_database_control_settings(dbcontrol_fixture):
+    """The minmum an URI should contain
+    """
+    fixture = dbcontrol_fixture
+    fixture.config.reahlsystem.connection_uri = 'myprefix:///'
+    with expected(ProgrammerError, test='Please specify a database name in reahlsystem.connection_uri'):
+        SystemControl(fixture.config)
+
+
+class PasswordScenarios(Fixture):
+    @scenario
+    def simple(self):
+        self.password = 'password'
+        self.expected_password = 'password'
+
+    @scenario
+    def complicated(self):
+        self.expected_password = 'p#=sword'
+        self.password = urllib_parse.quote(self.expected_password)
+
+    @scenario
+    def empty(self):
+        self.password = ''
+        self.expected_password = None
+
+
+@with_fixtures(DBControlFixture, PasswordScenarios)
+def test_database_control_complicated_passwords(dbcontrol_fixture, password_fixture):
```

Comment:
Remove scenarios, only test complicated scenario, with all parts encoded.

Rename fixture names to scenarios where applicable.